app/src/main/java/com/ubergeek42/WeechatAndroid/fragments/BufferFragment.java

```diff
@@ -351,14 +360,124 @@ public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
     ////////////////////////////////////////////////////////////////////////////////////////////////
 
     private boolean tcInProgress;
+    private boolean tcViaRelayInProgress = false;
+
+    @MainThread @SuppressLint("SetTextI18n") private void tryTabComplete() {
+        if (buffer == null) return;
+        // https://github.com/weechat/weechat/issues/1484
+        // Since WeeChat 2.9, relay clients can use the WeeChat completion mechanism.
+
+        // check for tcInProgress here if the previous tab complete ran
+        if (RelayConnection.weechatVersion >= 0x2090000 && !tcInProgress) {
+            tabCompleteRelayStrategy();
+        } else {
+            tabCompleteLocalStrategy();
+        }
+    }
+
+    private String tcViaRelayBaseString; // the entire original input
+    private int tcViaRelayIndex; // index used when looping over tcViaRelayCompletionResults
+    private String[] tcViaRelayCompletionResults;
+    private boolean tcViaRelayAddSpace; // to add a space after completion
+    private int tcViaRelayPosStart; // beginning of section to be replaced by one of the suggestions
+    private int tcViaRelayPosEnd; // end of said section in tcViaRelayBaseString
+    // tcViaRelayBaseString.subString(tcViaRelayPosStart, tcViaRelayPosEnd + 1)
+    private String tcViaRelayBaseWord;
+    private int tcViaRelaySelectionIndex; // where the cursor was when tab was pressed
+
+    @MainThread private void tabCompleteRelayStrategy() {
+        // disable the uiInput, re-enable it after the future has been completed
+        uiInput.setFocusable(false);
+
+        if (tcViaRelayInProgress) {
+            tcViaRelayIndex = (tcViaRelayIndex + 1) % tcViaRelayCompletionResults.length;
+            StringBuilder st = new StringBuilder(tcViaRelayBaseString);
+            st.replace(tcViaRelayPosStart, tcViaRelayPosEnd + 1,
+                    tcViaRelayCompletionResults[tcViaRelayIndex] +
+                            (tcViaRelayAddSpace ? " " : ""));
+            uiInput.setText(st.toString());
+            uiInput.setSelection(tcViaRelaySelectionIndex +
+                    tcViaRelayCompletionResults[tcViaRelayIndex].length()
+                    - tcViaRelayBaseWord.length() + (tcViaRelayAddSpace ? 1 : 0));
+            uiInput.setFocusable(true);
+            uiInput.setFocusableInTouchMode(true);
+            uiInput.setPressed(true);
+            tcViaRelayInProgress = true;
```

Comment:
can this be false here?

edit: actually yeah it will be false here! might want to put a comment here